File: test-scripts/keria-signify-delegation/signify-ts-runner/src/delegation.ts
import fs from 'node:fs';
import path from 'node:path';
import assert from 'node:assert/strict';
import {
    Algos,
    HabState,
    Operation,
    Siger,
    SignifyClient,
    Tier,
    assertMultisigIcp,
    d,
    messagize,
    ready,
} from 'signify-ts';

const ADMIN_URL = process.env.KERIA_ADMIN_URL ?? 'http://keria:3901';
const BOOT_URL = process.env.KERIA_BOOT_URL ?? 'http://keria:3903';
const QVI_DATA_DIR = process.env.QVI_DATA_DIR ?? '/qvi-data';
const SCENARIO = process.env.SCENARIO ?? 'signify-ts';
const ARTIFACT = path.join(QVI_DATA_DIR, `${SCENARIO}-delegation.json`);

const GEDA_PRE = process.env.GEDA_PRE;
const GEDA_OOBI = process.env.GEDA_OOBI;

const WIL_PRE = 'BLskRTInXnMxWaGqcpSyMgo0nYbalW99cGZESrz3zapM';
const WIL_OOBI = `http://witness-demo:5643/oobi/${WIL_PRE}/controller?name=Wil&tag=witness`;
const QVI_WITS = [WIL_PRE];
const QVI_NAME = 'signify-ts-qvi';
const MEMBERS: Array<[string, string]> = [
    ['signify-ts-qar1', 'tsqar1abcdefghijklmno'],
    ['signify-ts-qar2', 'tsqar2abcdefghijklmno'],
    ['signify-ts-qar3', 'tsqar3abcdefghijklmno'],
];

const POLL_INTERVAL_MS = Number(process.env.SIGNIFY_TS_DELEGATION_POLL_INTERVAL_MS ?? '500');
const DEFAULT_TIMEOUT_MS = Number(process.env.SIGNIFY_TS_DELEGATION_TIMEOUT_MS ?? '180000');

type AnyOperation = Operation | { name: string; done?: boolean; [key: string]: any };

interface Artifact {
    client: string;
    gedaPre: string;
    qviPre: string;
    members: string[];
    anchor: { i: string; s: string; d: string };
    operations: string[];
}

function log(message: string): void {
    console.log(`[signify-ts] ${message}`);
}

function requireEnv(): { gedaPre: string; gedaOobi: string } {
    assert.ok(GEDA_PRE, 'GEDA_PRE is required');
    assert.ok(GEDA_OOBI, 'GEDA_OOBI is required');
    return { gedaPre: GEDA_PRE, gedaOobi: GEDA_OOBI };
}

function sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
}

async function waitUntil<T>(
    fetcher: () => Promise<T>,
    readyFn: (value: T) => boolean,
    describe: string,
    timeoutMs = DEFAULT_TIMEOUT_MS
): Promise<T> {
    const deadline = Date.now() + timeoutMs;
    let lastValue: T | undefined;
    let lastError: unknown;
    while (true) {
        try {
            const value = await fetcher();
            lastValue = value;
            if (readyFn(value)) {
                return value;
            }
        } catch (error) {
            lastError = error;
        }

        if (Date.now() >= deadline) {
            throw new Error(
                `timed out waiting for ${describe}; lastError=${String(
                    lastError
                )}; lastValue=${JSON.stringify(lastValue)}`
            );
        }
        await sleep(POLL_INTERVAL_MS);
    }
}

async function connectClient(passcode: string): Promise<SignifyClient> {
    await ready();
    const client = new SignifyClient(ADMIN_URL, passcode.padEnd(21, '_'), Tier.low, BOOT_URL);
    try {
        await client.connect();
    } catch {
        const response = await client.boot();
        if (!response.ok) {
            throw new Error(`failed to boot SignifyTS client: ${response.status} ${await response.text()}`);
        }
        await client.connect();
    }
    return client;
}

async function clients(): Promise<SignifyClient[]> {
    return await Promise.all(MEMBERS.map(([, passcode]) => connectClient(passcode)));
}

async function waitOperation(
    client: SignifyClient,
    operation: AnyOperation | string,
    timeoutMs = DEFAULT_TIMEOUT_MS
): Promise<AnyOperation> {
    let op: AnyOperation =
        typeof operation === 'string'
            ? ((await client.operations().get(operation)) as AnyOperation)
            : operation;

    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), timeoutMs);
    try {
        op = (await client.operations().wait(op as Operation, {
            signal: controller.signal,
            minSleep: POLL_INTERVAL_MS,
            maxSleep: POLL_INTERVAL_MS,
            increaseFactor: 1,
        })) as AnyOperation;
    } finally {
        clearTimeout(timer);
    }

    if (op.name) {
        try {
            await client.operations().delete(op.name);
        } catch {
            // The operation may already have been removed by a dependency wait.
        }
    }
    return op;
}

async function resolveOobi(client: SignifyClient, oobi: string, alias: string): Promise<void> {
    await waitOperation(client, await client.oobis().resolve(oobi, alias));
}

async function ensureWitness(client: SignifyClient): Promise<void> {
    await resolveOobi(client, WIL_OOBI, 'wil');
}

async function waitForIdentifierOobi(
    client: SignifyClient,
    name: string,
    role: string
): Promise<string> {
    const oobis = await waitUntil(
        async () => (await client.oobis().get(name, role)).oobis,
        (value) => value.length > 0,
        `${role} OOBI for ${name}`
    );
    return oobis[0];
}

async function waitForEndRole(client: SignifyClient, name: string, eid: string): Promise<void> {
    const aid = await client.identifiers().get(name);
    await waitUntil(
        async () => await client.oobis().endroles(aid.prefix, 'agent'),
        (roles) => roles.some((role: any) => role.eid === eid),
        `agent end role for ${name}`
    );
}

async function getOrCreateIdentifier(client: SignifyClient, name: string): Promise<HabState> {
    try {
        return await client.identifiers().get(name);
    } catch {
        // Create below.
    }

    await ensureWitness(client);
    const result = await client.identifiers().create(name, {
        toad: 1,
        wits: QVI_WITS,
    });
    await waitOperation(client, await result.op());

    const endRoleResult = await client
        .identifiers()
        .addEndRole(name, 'agent', client.agent!.pre);
    await waitOperation(client, await endRoleResult.op());
    await waitForEndRole(client, name, client.agent!.pre);
    await waitForIdentifierOobi(client, name, 'agent');

    const aid = await client.identifiers().get(name);
    log(`created member ${name}: ${aid.prefix}`);
    return aid;
}

async function resolveAgentOobi(
    source: SignifyClient,
    sourceName: string,
    target: SignifyClient
): Promise<void> {
    const oobi = await waitForIdentifierOobi(source, sourceName, 'agent');
    await resolveOobi(target, oobi, sourceName);
}

async function exchangeAgentOobis(participants: Array<[SignifyClient, string]>): Promise<void> {
    for (let i = 0; i < participants.length; i += 1) {
        const [source, sourceName] = participants[i];
        for (const [target, targetName] of participants.slice(i + 1)) {
            await resolveAgentOobi(source, sourceName, target);
            await resolveAgentOobi(target, targetName, source);
        }
    }
}

function normalizeState(value: any): any {
    if (Array.isArray(value)) {
        assert.equal(value.length, 1);
        return value[0];
    }
    return value;
}

async function getStates(client: SignifyClient, prefixes: string[]): Promise<any[]> {
    const states = [];
    for (const prefix of prefixes) {
        states.push(normalizeState(await client.keyStates().get(prefix)));
    }
    return states;
}

function embedEvent(serder: any, sigs: string[]): [any, string] {
    const sigers = sigs.map((sig) => new Siger({ qb64: sig }));
    const ims = d(messagize(serder, sigers));
    return [serder, ims.substring(serder.size)];
}

async function unreadNotes(client: SignifyClient, route: string): Promise<any[]> {
    const response = await client.notifications().list();
    return response.notes.filter((note: any) => note.r === false && note.a?.r === route);
}

async function waitForMultisigRequest(client: SignifyClient, route: string): Promise<any[]> {
    const notes = await waitUntil(
        async () => await unreadNotes(client, route),
        (value) => value.length > 0,
        `${route} notification`
    );
    const note = notes[notes.length - 1];
    await client.notifications().mark(note.i);
    return await waitUntil(
        async () => await client.groups().getRequest(note.a.d),
        (request) => request.length > 0 && request[0].exn.r === route,
        `${route} request payload`
    );
}

async function startMultisigIncept(
    client: SignifyClient,
    input: {
        groupName: string;
        localMemberName: string;
        participants: string[];
        delpre: string;
    }
): Promise<{ operation: AnyOperation; serder: any }> {
    const member = await client.identifiers().get(input.localMemberName);
    const states = await getStates(client, input.participants);
    const result = await client.identifiers().create(input.groupName, {
        algo: Algos.group,
        mhab: member,
        isith: ['1/3', '1/3', '1/3'],
        nsith: ['1/3', '1/3', '1/3'],
        toad: 1,
        wits: QVI_WITS,
        delpre: input.delpre,
        states,
        rstates: states,
    });

    const serder = result.serder;
    const smids = states.map((state) => state.i);
    const recipients = input.participants.filter((prefix) => prefix !== member.prefix);
    await client.exchanges().send(
        input.localMemberName,
        'multisig',
        member,
        '/multisig/icp',
        { gid: serder.pre, smids, rmids: smids },
        { icp: embedEvent(serder, result.sigs) },
        recipients
    );
    return { operation: await result.op(), serder };
}

async function acceptMultisigIncept(
    client: SignifyClient,
    input: { groupName: string; localMemberName: string }
): Promise<AnyOperation> {
    const request = await waitForMultisigRequest(client, '/multisig/icp');
    const groupExn = assertMultisigIcp(request[0] as any) as any;
    const icp = groupExn.exn.e.icp;
    const smids = groupExn.exn.a.smids;
    const rmids = groupExn.exn.a.rmids ?? smids;
    const member = await client.identifiers().get(input.localMemberName);
    const states = await getStates(client, smids);
    const rstates = await getStates(client, rmids);
    const result = await client.identifiers().create(input.groupName, {
        algo: Algos.group,
        mhab: member,
        isith: icp.kt,
        nsith: icp.nt,
        toad: parseInt(icp.bt),
        wits: icp.b,
        delpre: icp.di,
        states,
        rstates,
    });
    const recipients = smids.filter((prefix: string) => prefix !== member.prefix);
    await client.exchanges().send(
        input.localMemberName,
        'multisig',
        member,
        '/multisig/icp',
        { gid: result.serder.pre, smids, rmids },
        { icp: embedEvent(result.serder, result.sigs) },
        recipients
    );
    return await result.op();
}

async function assertMultisigMembers(
    client: SignifyClient,
    groupName: string,
    expected: string[]
): Promise<void> {
    const members = await client.identifiers().members(groupName);
    assert.deepEqual(
        new Set((members as any).signing.map((member: any) => member.aid)),
        new Set(expected)
    );
    assert.deepEqual(
        new Set((members as any).rotation.map((member: any) => member.aid)),
        new Set(expected)
    );
}

async function waitForGroupState(
    client: SignifyClient,
    groupName: string,
    qviPre: string,
    gedaPre: string
): Promise<HabState> {
    return await waitUntil(
        async () => await client.identifiers().get(groupName),
        (group) =>
            group.prefix === qviPre &&
            group.state.di === gedaPre &&
            group.state.s === '0',
        `delegated group state for ${groupName}`
    );
}

async function clearGroupOperation(
    client: SignifyClient,
    operationName: string,
    groupName: string,
    qviPre: string,
    gedaPre: string
): Promise<void> {
    await waitForGroupState(client, groupName, qviPre, gedaPre);
    let operation: AnyOperation;
    try {
        operation = (await client.operations().get(operationName)) as AnyOperation;
    } catch {
        return;
    }

    if (operation.done === true) {
        await waitOperation(client, operation);
        return;
    }
    // KERIA 0.4.x can leave the group op pending after the identifier state is
    // materialized. The state check above is the convergence proof for this test.
    await client.operations().delete(operationName);
}

async function unfinishedOperations(client: SignifyClient): Promise<AnyOperation[]> {
    const operations = (await client.operations().list()) as AnyOperation[];
    return operations.filter((operation) => operation.done === false);
}

async function setup(): Promise<void> {
    const { gedaPre, gedaOobi } = requireEnv();
    fs.mkdirSync(QVI_DATA_DIR, { recursive: true });
    const cs = await clients();
    const names = MEMBERS.map(([name]) => name);

    const aids = [];
    for (const [client, name] of cs.map((client, index) => [client, names[index]] as const)) {
        aids.push(await getOrCreateIdentifier(client, name));
    }
    await exchangeAgentOobis(cs.map((client, index) => [client, names[index]] as [SignifyClient, string]));
    for (const client of cs) {
        await resolveOobi(client, gedaOobi, 'geda');
    }

    const participants = aids.map((aid) => aid.prefix);
    const { operation, serder } = await startMultisigIncept(cs[0], {
        groupName: QVI_NAME,
        localMemberName: names[0],
        participants,
        delpre: gedaPre,
    });
    const operations = [operation];
    for (let i = 1; i < cs.length; i += 1) {
        operations.push(
            await acceptMultisigIncept(cs[i], {
                groupName: QVI_NAME,
                localMemberName: names[i],
            })
        );
    }

    const artifact: Artifact = {
        client: 'signify-ts',
        gedaPre,
        qviPre: serder.pre,
        members: participants,
        anchor: { i: serder.pre, s: '0', d: serder.pre },
        operations: operations.map((op) => op.name),
    };
    fs.writeFileSync(ARTIFACT, `${JSON.stringify(artifact, null, 2)}\n`);
    log(`wrote delegated inception artifact for ${serder.pre}`);
}

async function complete(): Promise<void> {
    assert.ok(fs.existsSync(ARTIFACT), `missing artifact ${ARTIFACT}`);
    const artifact = JSON.parse(fs.readFileSync(ARTIFACT, 'utf8')) as Artifact;
    const cs = await clients();
    assert.deepEqual(artifact.anchor, { i: artifact.qviPre, s: '0', d: artifact.qviPre });

    for (const client of cs) {
        const result = await waitOperation(
            client,
            await client.keyStates().query(artifact.gedaPre, '1')
        );
        const state = normalizeState((result as any).response);
        assert.equal(state.i, artifact.gedaPre);
        assert.ok(Number.parseInt(state.s, 10) >= 1);
    }
    for (let i = 0; i < cs.length; i += 1) {
        await clearGroupOperation(
            cs[i],
            artifact.operations[i],
            QVI_NAME,
            artifact.qviPre,
            artifact.gedaPre
        );
    }

    const groups = [];
    for (const client of cs) {
        groups.push(await client.identifiers().get(QVI_NAME));
    }
    assert.deepEqual(new Set(groups.map((group) => group.prefix)), new Set([artifact.qviPre]));
    for (const group of groups) {
        assert.equal(group.state.di, artifact.gedaPre);
        assert.equal(group.state.s, '0');
    }

    await assertMultisigMembers(cs[0], QVI_NAME, artifact.members);
    const incomplete = [];
    for (const client of cs) {
        incomplete.push(...(await unfinishedOperations(client)));
    }
    assert.equal(incomplete.length, 0, `incomplete operations remain: ${JSON.stringify(incomplete)}`);
    log('PASS: SignifyTS multisig delegate approved by KERIpy multisig KLI delegator');
}

async function main(): Promise<void> {
    const command = process.argv[2];
    if (command === 'setup') {
        await setup();
    } else if (command === 'complete') {
        await complete();
    } else {
        throw new Error('usage: tsx src/delegation.ts setup|complete');
    }
}

main().catch((error) => {
    console.error(error);
    process.exit(1);
});

File: test-scripts/keria-signify-delegation/signifypy-runner/src/delegation.py
import json
import os
import sys
import time
from pathlib import Path

from keri.app.keeping import Algos
from keri.core import eventing, serdering
from keri.core import signing as csigning
from keri.core.coring import Tiers
from requests import HTTPError
from signify.app.clienting import SignifyClient


ADMIN_URL = os.environ.get("KERIA_ADMIN_URL", "http://keria:3901")
BOOT_URL = os.environ.get("KERIA_BOOT_URL", "http://keria:3903")
ARTIFACT = Path(os.environ.get("QVI_DATA_DIR", "/qvi-data")) / "signifypy-delegation.json"

WIL_PRE = "BLskRTInXnMxWaGqcpSyMgo0nYbalW99cGZESrz3zapM"
WIL_OOBI = f"http://witness-demo:5643/oobi/{WIL_PRE}/controller?name=Wil&tag=witness"
QVI_WITS = [WIL_PRE]
QVI_THRESHOLD = ["1/2", "1/2", "1/2"]
QVI_NAME = "signifypy-qvi"
MEMBERS = [
    ("signifypy-qar1", "pyqar1abcdefghijklmno"),
    ("signifypy-qar2", "pyqar2abcdefghijklmno"),
    ("signifypy-qar3", "pyqar3abcdefghijklmno"),
]

POLL_INTERVAL = float(os.environ.get("SIGNIFYPY_DELEGATION_POLL_INTERVAL", "0.5"))
DEFAULT_TIMEOUT = float(os.environ.get("SIGNIFYPY_DELEGATION_TIMEOUT", "180"))


def log(message: str) -> None:
    print(f"[signifypy] {message}", flush=True)


def wait_until(fetch, ready, *, describe: str, timeout: float = DEFAULT_TIMEOUT):
    deadline = time.monotonic() + timeout
    last_value = None
    last_error: str | None = None
    while True:
        try:
            value = fetch()
        except HTTPError as err:
            last_error = str(err)
        else:
            last_value = value
            if ready(value):
                return value

        if time.monotonic() >= deadline:
            raise TimeoutError(
                f"timed out waiting for {describe}; "
                f"last_error={last_error!r}; last_value={last_value!r}"
            )
        time.sleep(POLL_INTERVAL)


def connect_client(passcode: str) -> SignifyClient:
    client = SignifyClient(
        passcode=passcode,
        tier=Tiers.low,
        url=ADMIN_URL,
        boot_url=BOOT_URL,
    )
    try:
        client.connect()
    except Exception:
        client.boot()
        client.connect()
    return client


def clients() -> list[SignifyClient]:
    return [connect_client(passcode) for _, passcode in MEMBERS]


def wait_operation(client: SignifyClient, operation: dict | str, *, timeout: float = DEFAULT_TIMEOUT) -> dict:
    if isinstance(operation, str):
        operation = client.operations().get(operation)
    result = client.operations().wait(
        operation,
        timeout=timeout,
        interval=POLL_INTERVAL,
        max_interval=POLL_INTERVAL,
        backoff=1.0,
    )
    return result


def resolve_oobi(client: SignifyClient, oobi: str, alias: str) -> None:
    wait_operation(client, client.oobis().resolve(oobi, alias=alias))


def wait_for_multisig_request(client: SignifyClient, route: str) -> list[dict]:
    note = wait_until(
        lambda: [
            note
            for note in client.notifications().list()["notes"]
            if note.get("r") is False and note.get("a", {}).get("r") == route
        ],
        bool,
        describe=f"{route} notification",
    )[-1]
    client.notifications().mark(note["i"])
    return wait_until(
        lambda: client.groups().get_request(note["a"]["d"]),
        lambda request: bool(request) and request[0]["exn"]["r"] == route,
        describe=f"{route} request payload",
    )


def normalize_state(value):
    if isinstance(value, list):
        if len(value) != 1:
            raise AssertionError(f"expected one key state, got {len(value)}")
        return value[0]
    return value


def get_states(client: SignifyClient, prefixes: list[str]) -> list[dict]:
    return [normalize_state(client.keyStates().get(prefix)) for prefix in prefixes]


def wait_for_identifier_oobi(client: SignifyClient, name: str) -> str:
    return wait_until(
        lambda: client.oobis().get(name, role="agent")["oobis"],
        bool,
        describe=f"agent OOBI for {name}",
    )[0]


def create_identifier(client: SignifyClient, name: str) -> dict:
    resolve_oobi(client, WIL_OOBI, "wil")
    _, _, operation = client.identifiers().create(name, toad="1", wits=QVI_WITS)
    wait_operation(client, operation)
    _, _, operation = client.identifiers().addEndRole(name=name, eid=client.agent.pre)
    wait_operation(client, operation)
    wait_for_identifier_oobi(client, name)
    aid = client.identifiers().get(name)
    log(f"created member {name}: {aid['prefix']}")
    return aid


def exchange_agent_oobis(participants: list[tuple[SignifyClient, str]]) -> None:
    for i, (source, source_name) in enumerate(participants):
        for target, target_name in participants[i + 1 :]:
            resolve_oobi(target, wait_for_identifier_oobi(source, source_name), source_name)
            resolve_oobi(source, wait_for_identifier_oobi(target, target_name), target_name)


def messagize(serder, sigs):
    sigers = [csigning.Siger(qb64=sig) for sig in sigs]
    return eventing.messagize(serder=serder, sigers=sigers)


def start_multisig_incept(
    client: SignifyClient, local_member_name: str, participants: list[str], delpre: str
) -> serdering.SerderKERI:
    member = client.identifiers().get(local_member_name)
    states = get_states(client, participants)
    serder, sigs, _ = client.identifiers().create(
        QVI_NAME,
        algo=Algos.group,
        mhab=member,
        isith=QVI_THRESHOLD,
        nsith=QVI_THRESHOLD,
        toad=1,
        wits=QVI_WITS,
        delpre=delpre,
        states=states,
        rstates=states,
    )
    smids = [state["i"] for state in states]
    client.exchanges().send(
        local_member_name,
        "multisig",
        sender=member,
        route="/multisig/icp",
        payload={"gid": serder.pre, "smids": smids, "rmids": smids},
        embeds={"icp": messagize(serder, sigs)},
        recipients=[prefix for prefix in participants if prefix != member["prefix"]],
    )
    return serder


def accept_multisig_incept(client: SignifyClient, local_member_name: str) -> None:
    request = wait_for_multisig_request(client, "/multisig/icp")
    exn = request[0]["exn"]
    icp = exn["e"]["icp"]
    smids = exn["a"]["smids"]
    rmids = exn["a"].get("rmids", smids)
    member = client.identifiers().get(local_member_name)
    serder, sigs, _ = client.identifiers().create(
        QVI_NAME,
        algo=Algos.group,
        mhab=member,
        isith=icp["kt"],
        nsith=icp["nt"],
        toad=int(icp["bt"]),
        wits=icp["b"],
        delpre=icp.get("di"),
        states=get_states(client, smids),
        rstates=get_states(client, rmids),
    )
    client.exchanges().send(
        local_member_name,
        "multisig",
        sender=member,
        route="/multisig/icp",
        payload={"gid": serder.pre, "smids": smids, "rmids": rmids},
        embeds={"icp": messagize(serder, sigs)},
        recipients=[prefix for prefix in smids if prefix != member["prefix"]],
    )


def wait_for_group_state(client: SignifyClient, qvi_pre: str, geda_pre: str) -> dict:
    return wait_until(
        lambda: client.identifiers().get(QVI_NAME),
        lambda group: (
            group["prefix"] == qvi_pre
            and group["state"]["di"] == geda_pre
            and group["state"]["s"] == "0"
            and group["state"]["kt"] == QVI_THRESHOLD
            and group["state"]["nt"] == QVI_THRESHOLD
        ),
        describe=f"2-of-3 delegated group state for {QVI_NAME}",
    )


def assert_multisig_members(client: SignifyClient, expected: list[str]) -> None:
    members = client.identifiers().members(QVI_NAME)
    for role in ("signing", "rotation"):
        actual = {member["aid"] for member in members[role]}
        if actual != set(expected):
            raise AssertionError(f"unexpected {role} members: {actual}, expected={set(expected)}")


def setup() -> None:
    geda_pre = os.environ.get("GEDA_PRE")
    geda_oobi = os.environ.get("GEDA_OOBI")
    if not geda_pre or not geda_oobi:
        raise RuntimeError("GEDA_PRE and GEDA_OOBI are required")
    cs = clients()
    names = [name for name, _ in MEMBERS]
    aids = [create_identifier(client, name) for client, name in zip(cs, names)]
    exchange_agent_oobis(list(zip(cs, names)))
    for client in cs:
        resolve_oobi(client, geda_oobi, "geda")

    participants = [aid["prefix"] for aid in aids]
    serder = start_multisig_incept(cs[0], names[0], participants, geda_pre)
    for client, name in zip(cs[1:], names[1:]):
        accept_multisig_incept(client, name)

    artifact = {"gedaPre": geda_pre, "qviPre": serder.pre, "members": participants}
    ARTIFACT.write_text(json.dumps(artifact, indent=2, sort_keys=True) + "\n")
    log(f"wrote delegated inception artifact for {serder.pre}")


def complete() -> None:
    if not ARTIFACT.exists():
        raise RuntimeError(f"missing artifact {ARTIFACT}")
    artifact = json.loads(ARTIFACT.read_text())
    cs = clients()
    geda_pre = artifact["gedaPre"]
    qvi_pre = artifact["qviPre"]

    for client in cs:
        state = normalize_state(
            wait_operation(client, client.keyStates().query(pre=geda_pre, sn="1"))["response"]
        )
        if state["i"] != geda_pre or int(state["s"]) < 1:
            raise AssertionError(f"GEDA query did not converge on approval interaction: {state}")

    groups = [wait_for_group_state(client, qvi_pre, geda_pre) for client in cs]
    for (name, _), group in zip(MEMBERS, groups):
        state = group["state"]
        log(
            f"verified {name}: prefix={group['prefix']} di={state['di']} "
            f"s={state['s']} kt={state['kt']} nt={state['nt']}"
        )
    assert_multisig_members(cs[0], artifact["members"])
    log("PASS: SignifyPy 2-of-3 multisig delegate approved by KERIpy multisig KLI delegator")


def main() -> None:
    if len(sys.argv) != 2 or sys.argv[1] not in {"setup", "complete"}:
        raise SystemExit("usage: delegation.py setup|complete")
    setup() if sys.argv[1] == "setup" else complete()


if __name__ == "__main__":
    main()
